q2_longitudinal/_utilities.py

```diff
@@ -577,3 +578,74 @@ def _stats_and_visuals(output_dir, pairs, metric, group_column,
                paired_difference_tests, boxplot, summary=summary,
                errors=errors, plot_name=plot_name,
                pairwise_test_name=pairwise_test_name)
+
+
+def _temporal_corr(table, individual_id, id_set, corr_method="kendall"):
+    '''Create Temporal correlation from a feature table containing repeated
+    measures samples.
+    table: pd.DataFrame
+        rows are samples, columns are features (count / relative_abundance)
+    individual_id: pd.Series
+        subject id of samples, with the same length as df
+    id_set: pd.Series
+        unique subject ids from individual_id with index attached
+    corr_method: str
+        temporal correlation method, "kendall", "pearson", "spearman"
+
+    '''
+    results = {}
+
+    # Start to calculate temporal correlation
+    for id_key in id_set:
+        table_id = table.loc[individual_id == id_key]
+        results[id_key] = table_id.corr(method=corr_method).fillna(0)
+
+    return results
+
+
+def _temporal_distance(corr, id_set, dist_method="fro"):
+    '''Calculate Distance Matrix from temporal correlation data.
+    corr_dict: dict
+        output from temporal_corr
+    id_set: pd.Series
+        unique subject ids from individual_id with index attached
+    dist_method: str
+        method supported by scipy.linalg.norm parameter ord
+    '''
+    id_n = len(id_set)
+
+    dist = np.zeros((id_n, id_n))
+    for i, id_i in enumerate(id_set):
```

Comment:
@elong0527 it looks like there may be two unnecessary steps in this loop:
1. this will calculate the distance of each `(id_i, id_j)` pair twice. we should instead just duplicate the result if this creates a symmetric matrix
2. this calculates the identical-sample distance along the diagonal of the matrix. Should this just be skipped so that all self-comparisons are 0?

Reply:
Yes, you are right. I will update the code to improve performance.